### PR TITLE
In-place addition, multiplication, subtraction of usm_ndarrays

### LINKED ISSUE #1229: Implement in place functions
Implement call without extra copy for functions: add, divide, multiply, subtract.

``` python
dpt.add(a, b, out=a)

# should work like
a += b

# not like
a[...] = a + b
```

### PR DESCRIPTION
**Pull request by ndgrigorian**
Closes #1229 
This pull request implements in-place addition, subtraction, and multiplication of ``usm_ndarray`` instances.

By avoiding the additional allocation of an output array, the performance advantage compared compared to applying the operator with a scalar, a la ``dpt.add(x, 2)``, is considerable.

For example:
```
In [8]: X = dpt.ones((16768, 16768), dtype="f4")

In [9]: %time X += 2
CPU times: user 58.5 ms, sys: 78.1 ms, total: 137 ms
Wall time: 133 ms

In [10]: %time dpt.add(X, 2)
CPU times: user 238 ms, sys: 147 ms, total: 385 ms
Wall time: 392 ms
```

- [X] Have you provided a meaningful PR description?
- [X] Have you added a test, reproducer or referred to an issue with a reproducer?
- [X] Have you tested your changes locally for CPU and GPU devices?
- [X] Have you made sure that new changes do not introduce compiler warnings?
- [X] Have you checked performance impact of proposed changes?
- [X] If this PR is a work in progress, are you opening the PR as a draft?

### CODE REVIEW

File: dpctl/tensor/_elementwise_common.py [hunk 1]
```diff
@@ -631,3 +640,115 @@ def __call__(self, o1, o2, out=None, order="K"):
...
+            lhs_dtype, val_dtype, sycl_dev
+        )
+
+        buf_dt = _find_inplace_dtype(
+            lhs_dtype, val_dtype, self.result_type_resolver_fn_, sycl_dev
+        )
+
+        if buf_dt is None:
+            raise TypeError(
+                f"function '{self.name_}' does not support input types "
```
[the remark this answers is not in the document]
Comment (ndgrigorian) [reply dpctl/tensor/_elementwise_common.py:718]: After considering it, binary operations between dtype 'kinds' are currently only supported for output of the default dtype for a resulting kind. In the case of complex and integer, it requires complex128.

So for now, I won't change this for in-place operations to remain consistent with other binary operations.

File: dpctl/tensor/libtensor/include/kernels/elementwise_functions/common_inplace.hpp [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+
+        const auto &inp_offset = two_offsets_.get_first_offset();
+        const auto &lhs_offset = two_offsets_.get_second_offset();
+
+        BinaryInplaceOperatorT op{};
+        op(lhs[lhs_offset], rhs[inp_offset]);
+    }
+};
+
+// Typdefs for function pointers
```
Comment (oleksandr-pavlyk) [inline dpctl/tensor/libtensor/include/kernels/elementwise_functions/common_inplace.hpp:194]: Typdfs -> Typedefs

**Review by vtavana — APPROVED** [review]
Thank you, @ndgrigorian

**Review by AlexanderKalistratov — COMMENTED** [review]
Just small comments. You can ignore it, if you want.

**Pull request merged**